src/ng_package/text_replace/index.ts

```diff
@@ -0,0 +1,64 @@
+import fs from 'node:fs/promises';
+import path from 'node:path';
+import {globSync} from 'tinyglobby';
+
+/**
+ * Interface describing a file captured in the Bazel action.
+ * https://docs.bazel.build/versions/main/skylark/lib/File.html.
+ */
+export interface BazelFileInfo {
+  /** Execroot-relative path pointing to the file. */
+  path: string;
+  /** The path of this file relative to its root. e.g. omitting `bazel-out/<..>/bin`. */
+  shortPath: string;
+}
+
+/** Apply the provided substitutions to the content of the file at the origin, writing it to the provided destination. */
+async function applySubstitutions(origin: string, destination: string, substitutions: [RegExp, string][]) {
+  // Ensure that a directory exists for the destination file to be written into. We don't guarantee order of substitutions,
+  // so we need to make sure manually for the provided file paths.
+  await fs.mkdir(path.dirname(destination), {recursive: true});
+
+  /** The content of the file to the run the substitutions against. */
+  let content = await fs.readFile(origin, {encoding: 'utf-8'});  
+
+  substitutions.forEach(([regexp, replacement]) => {
+    content = content.replace(regexp, replacement);
+  });
+
+  return fs.writeFile(destination, content);
+}
+
+async function main(args: string[]) {
+  const [substitutionsArg, originsRaw, destinationRaw] = (await fs.readFile(args[0], {encoding: 'utf-8'})).split('\n').map(line => line.replace(/^'(.*)'$/, '$1'));
+  /** List of files to propcess for substitution */
+  const origins = JSON.parse(originsRaw) as BazelFileInfo[];
+  /** The destination directory for the copied files with substitutions applied. */
+  const destinationDir = JSON.parse(destinationRaw) as BazelFileInfo;
+  /** Map of substitutions to make whenever the regex key is matched in a file's content. */
+  const substitutions = Object
+    .entries(JSON.parse(substitutionsArg) as Record<string, string>)
+    .map<[RegExp, string]>(([key, val]) => ([new RegExp(key, 'g'), val]));
+  
+  /** Discovered file paths to apply substitutions to, split into the origin path and the file path based on that origin. */
+  let files: [string, string][] = [];
+
+  for (let origin of origins) {
+    files.push(
+      // Find all of the files in the origin directory and add them to the list.
+      ...globSync("**", {cwd: origin.path})
```

Comment:
This could be parallel too, but probably doesn't matter a lot with few origins.

Reply:
Does it make sense to just take a single directory? that would also make the output layout more obvious IMO